haystack/schema.py

```diff
@@ -550,36 +617,305 @@ def __setitem__(self, key: str, value: pd.DataFrame):
     def __contains__(self, key: str):
         return self.node_results.keys().__contains__(key)
 
-    def append(self, key: str, value: pd.DataFrame):
-        if key in self.node_results:
-            self.node_results[key] = pd.concat([self.node_results[key], value])
-        else:    
-            self.node_results[key] = value
+    def __len__(self):
+        return self.node_results.__len__()
 
-    def calculate_metrics(self) -> Dict[str, float]:
+    def append(self, key: str, value: pd.DataFrame):
+        if value is not None and len(value) > 0:
+            if key in self.node_results:
+                self.node_results[key] = pd.concat([self.node_results[key], value])
+            else:    
+                self.node_results[key] = value
+
+    def calculate_metrics(
+        self, 
+        simulated_top_k_reader: int = -1,
+        simulated_top_k_retriever: int = -1,
+        doc_relevance_col: str = "gold_id_match"
+    ) -> Dict[str, Dict[str, float]]:
         """
-            First dummy implementation of metrics calcuation just to show the way it's done.
-            TODO: implement retriever and reader specific metrics that must not rely on node names.
+        Calculates proper metrics for each node.
+
+        For document returning nodes default metrics are: 
+        - mrr (Mean Reciprocal Rank: see https://en.wikipedia.org/wiki/Mean_reciprocal_rank)
+        - map (Mean Average Precision: see https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Mean_average_precision)
+        - precision (Precision: How much of the returned documents were relevant?)
+        - recall_ir (Recall according to Information Retrieval definition: How much of the relevant documents were retrieved per query?)
+        - recall_qs (Recall for Question Answering: How much of the queries returned at least one relevant document?)
+
+        For answer returning nodes default metrics are:
+        - exact_match (How much of the queries returned the exact answer?)
+        - f1 (How well do the returned results overlap with any gold answer on token basis?)
+        - sas if a SAS model has bin provided during during pipeline.eval() (How well do the returned results overlap with any gold answer on a semantic basis?)
```

Comment:
Instead of `How well do the returned results overlap with any gold answer on a semantic basis?` I'd prefer something like `How semantically similar is the prediction to the gold answers?`. Happy to discuss

Reply:
done